src/flint/types/fmpq_mpoly.pyx

```diff
@@ -576,103 +577,89 @@ cdef class fmpq_mpoly(flint_mpoly):
     def __rmod__(self, other):
         return divmod(other, self)[1]
 
-    def __call__(self, *args, **kwargs):
+    def __call__(self, *args) -> fmpq:
+        cdef:
+            fmpq_vec V
+            fmpq vres
+            slong nvars = self.ctx.nvars(), nargs = len(args)
+
+        if nargs < nvars:
+            raise ValueError("Not enough arguments provided")
+        elif nargs > nvars:
+            raise ValueError("More arguments provided than variables")
+
+        args_fmpq = tuple(any_as_fmpq(v) for v in args)
+        for arg in args_fmpq:
+            if arg is NotImplemented:
+                raise TypeError(f"Cannot coerce argument ('{arg}') to fmpq")
+
+        V = fmpq_vec(args_fmpq, double_indirect=True)
+        vres = fmpq.__new__(fmpq)
+        if fmpq_mpoly_evaluate_all_fmpq(vres.val, self.val, V.double_indirect, self.ctx.val) == 0:
+            raise ValueError("Unreasonably large polynomial")  # pragma: no cover
+        return vres
+
+    def subs(self, dict_args) -> fmpq_mpoly:
+        """
+        Partial evaluate this polynomial.
+        """
         cdef:
             fmpq_mpoly res
             fmpq_mpoly res2
+            slong i, nargs
+
+        partial_args = tuple((i, dict_args[x]) for i, x in enumerate(self.ctx.names()) if x in dict_args)
+        nargs = len(partial_args)
+
+        # If we've been provided with an invalid keyword arg then the length of our filter
+        # args will be less than what we've been provided with.
+        # If the length is equal to the number of variables then all arguments have been provided
+        # otherwise we need to do partial application
+        if nargs < len(dict_args):
+            raise ValueError("Unknown keyword argument provided")
+
+        args_fmpq = tuple(any_as_fmpq(v) for _, v in partial_args)
+        for arg in args_fmpq:
+            if arg is NotImplemented:
+                raise TypeError(f"Cannot coerce argument ('{arg}') to fmpq")
+
+        # Partial application with args in Z. We evaluate the polynomial one variable at a time
+        res = create_fmpq_mpoly(self.ctx)
+        res2 = create_fmpq_mpoly(self.ctx)
+
+        fmpq_mpoly_set(res2.val, self.val, self.ctx.val)
+        for (i, _), arg in zip(partial_args, args_fmpq):
+            if fmpq_mpoly_evaluate_one_fmpq(res.val, res2.val, i, (<fmpq>arg).val, self.ctx.val) == 0:
+                raise ValueError("Unreasonably large polynomial")  # pragma: no cover
+            fmpq_mpoly_set(res2.val, res.val, self.ctx.val)
```

Comment:
Does `fmpq_mpoly_evaluate_one_fmpq` support aliasing so that you can do
```
fmpq_mpoly_evaluate_one_fmpq(res.val, res.val, ...)
```

Reply:
> Does `fmpq_mpoly_evaluate_one_fmpq` support aliasing so that you can do
> 
> ```
> fmpq_mpoly_evaluate_one_fmpq(res.val, res.val, ...)
> ```

It appears it can. I'll make that change.
```c
int fmpq_mpoly_evaluate_one_fmpq(fmpq_mpoly_t A,
                           const fmpq_mpoly_t B, slong var, const fmpq_t val,
                                                   const fmpq_mpoly_ctx_t ctx)
{
...
    if (A == B)
    {
        int success;
        fmpq_mpoly_t T;
        fmpq_mpoly_init(T, ctx);
        success = fmpq_mpoly_evaluate_one_fmpq(T, B, var, val, ctx);
        fmpq_mpoly_swap(A, T, ctx);
        fmpq_mpoly_clear(T, ctx);
        return success;
    }
...
}
```